<a href="https://colab.research.google.com/github/ErickDiaz17799/ProyectosMaestria/blob/main/Desambiguaci%C3%B3n_del_sentido_de_las_palabras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Universidad Internacional de La Rioja (UNIR) - Máster Universitario en Inteligencia Artificial - Procesamiento del Lenguaje Natural**

***
Datos del alumno:
Díaz Tapia Erick Ulises

Fecha:
 03/08/2025

<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">Actividad 3: Desambiguación del sentido de las palabras</span>

**Objetivos**

Con este laboratorio el alumno conseguirá aplicar diferentes algoritmos basados en aprendizaje automático supervisado para desambiguar el sentido de las palabras. Además, va a aprender a utilizar la herramienta de software abierto Natural Language Toolkit (NLTK) con la que implementar tareas de procesamiento del lenguaje natural en Python.

**Descripción**
En este laboratorio debes desarrollar e implementar diferentes algoritmos basados en aprendizaje automático supervisado para desambiguar el sentido de las palabras en Python y utilizando la herramienta de software abierto Natural Language Toolkit (NLTK).

Para preparar este laboratorio, simplemente descarga e instala **NLTK 3.3** en tu equipo.

***
Accede a NLTK 3.3 desde la siguiente dirección web:
https://www.nltk.org/install.html
***

NLTK requiere de Python versiones 2.7, 3.4, 3.5 o 3.6 para funcionar. Por lo que, si no tienes instalado Python, descárgalo e instálalo.

***
Accede a **Python**  desde la siguiente dirección web: https://www.python.org/downloads/
***

Asegúrate de que has instalado NLTK 3.3 adecuadamente antes de la sesión de laboratorio y de revisar el contenido teórico, Semántica léxica y temas anteriores, para tener frescos los diferentes conceptos sobre el procesamiento del lenguaje natural estudiados en esta asignatura.

Durante la sesión del laboratorio debes solucionar un problema sobre desambiguación del sentido de las palabras utilizando el corpus etiquetado en inglés llamado **Senseval 2** y que viene disponible en NLTK.

***
Accede a más información sobre **Senseval 2** desde la siguiente dirección web:
http://www.nltk.org/howto/corpus.html
***

El primer paso es importar el corpus etiquetado utilizando los siguientes comandos:

In [206]:
import nltk
from nltk.corpus import senseval

El corpus **Senseval 2** contiene datos etiquetados que sirven para entrenar un clasificador que permita desambiguar el sentido de las palabras. Cada elemento del corpus **Senseval 2** se corresponde a una palabra ambigua. Concretamente en este laboratorio se trabajará con las palabras en inglés *«hard»* y *«serve»*, aunque en el corpus hay información de otras dos.

Para poder extraer la información sobre las palabras es imprescindible la manera en la que se identifican en el corpus, es decir, sus identificadores. Con el siguiente comando, se extraen los identificadores de las palabras tratadas en el corpus.

In [207]:
senseval.fileids()

['hard.pos', 'interest.pos', 'line.pos', 'serve.pos']

Para cada una de las palabras ambiguas, el corpus contiene una lista de instancias correspondientes a las ocurrencias de esa palabra. Para cada instancia se proporciona la palabra, una lista de sentidos que se aplican a la aparición de esa palabra y el contexto de la palabra.

En la siguiente figura se observa el comando utilizado para visualizar la información que contiene cada instancia de la palabra ambigua *«hard»*.

In [208]:
senseval.instances('hard.pos')

[SensevalInstance(word='hard-a', position=20, context=[('``', '``'), ('he', 'PRP'), ('may', 'MD'), ('lose', 'VB'), ('all', 'DT'), ('popular', 'JJ'), ('support', 'NN'), (',', ','), ('but', 'CC'), ('someone', 'NN'), ('has', 'VBZ'), ('to', 'TO'), ('kill', 'VB'), ('him', 'PRP'), ('to', 'TO'), ('defeat', 'VB'), ('him', 'PRP'), ('and', 'CC'), ('that', 'DT'), ("'s", 'VBZ'), ('hard', 'JJ'), ('to', 'TO'), ('do', 'VB'), ('.', '.'), ("''", "''")], senses=('HARD1',)), SensevalInstance(word='hard-a', position=10, context=[('clever', 'NNP'), ('white', 'NNP'), ('house', 'NNP'), ('``', '``'), ('spin', 'VB'), ('doctors', 'NNS'), ("''", "''"), ('are', 'VBP'), ('having', 'VBG'), ('a', 'DT'), ('hard', 'JJ'), ('time', 'NN'), ('helping', 'VBG'), ('president', 'NNP'), ('bush', 'NNP'), ('explain', 'VB'), ('away', 'RB'), ('the', 'DT'), ('economic', 'JJ'), ('bashing', 'NN'), ('that', 'IN'), ('low-and', 'JJ'), ('middle-income', 'JJ'), ('workers', 'NNS'), ('are', 'VBP'), ('taking', 'VBG'), ('these', 'DT'), ('days

In [209]:
# Cargar las instancias de "hard"
instances = senseval.instances('hard.pos')

# Imprimir las primeras 5 instancias con formato amigable
for i, instance in enumerate(instances[:5]):
    print(f"\nInstancia {i + 1}")

    # El contexto es una lista de tuplas (palabra, etiqueta gramatical)
    contexto = ' '.join([w for (w, _) in instance.context])
    print(f"Contexto: {contexto}")

    # Palabra objetivo
    print(f"Palabra objetivo: {instance.word}")

    # Sentidos
    print(f"Sentidos: {', '.join(instance.senses)}")

    print("-" * 50)



Instancia 1
Contexto: `` he may lose all popular support , but someone has to kill him to defeat him and that 's hard to do . ''
Palabra objetivo: hard-a
Sentidos: HARD1
--------------------------------------------------

Instancia 2
Contexto: clever white house `` spin doctors '' are having a hard time helping president bush explain away the economic bashing that low-and middle-income workers are taking these days .
Palabra objetivo: hard-a
Sentidos: HARD1
--------------------------------------------------

Instancia 3
Contexto: i find it hard to believe that the sacramento river will ever be quite the same , although i certainly wish that i 'm wrong .
Palabra objetivo: hard-a
Sentidos: HARD1
--------------------------------------------------

Instancia 4
Contexto: now when you get bad credit data or are confused with another person , the hard part in correcting the mistake is not even knowing where it is recorded , let alone having access .
Palabra objetivo: hard-a
Sentidos: HARD1
-

Por ejemplo, en la primera instancia (`SensevalInstance`) la palabra ambigua (`word`) es `‘hard-a’`, lo que indica que la palabra es `‘hard’` y en este caso la categoría gramatical es un adjetivo, identificado por el sufijo `‘-a’`.

El campo `position` indica la posición en la oración en la que se encuentra la palabra ambigua, en este caso la palabra `‘hard’` se encuentra en la posición 20.

El campo `context` representa el contexto, es decir, la oración en la que se encuentra la palabra ambigua, en este ejemplo *«"he may lose all popular support, but someone has to kill him to defeat him and that's hard to do."»*. El contexto viene representado por pares formados por una palabra y la correspondiente etiqueta gramatical. Por ejemplo, el par `(‘he’, ‘PRP’)` que aparece en el contexto indica que la categoría gramatical asociada a la palabra `‘he’` es un pronombre personal `‘PRP’`.

Por último, el campo `senses` contiene los posibles sentidos de la palabra ambigua, en el ejemplo `‘HARD1’`. Los sentidos del corpus hacen referencia a los sentidos de la palabra recogidos en la base de datos de relaciones léxicas WordNet<sup>1</sup>.


<sup>1</sup> Puede que los sentidos que aparecen en Senseval 2 difieran de los que se encuentran actualmente en WordNet, debido a la constante actualización de este. En este laboratorio no será necesario trabajar con WordNet, se menciona como información adicional.

En este caso `‘HARD1’` hace referencia la primera definición de la palabra `‘hard’` que aparece en WordNet, a «difícil»,  *«difficult, hard (not easy; requiring great physical or mental effort to accomplish or comprehend or endure)»*. Esta información se puede obtener utilizando la interfaz de búsqueda web de WordNet cuyo resultado se muestra en la siguiente figura.

***
Accede al interfaz de búsqueda de WordNet desde la siguiente dirección web: http://wordnetweb.princeton.edu/perl/webwn

***



**Nota:** NLTK implementa también un lector para la información disponible en la base de datos de relaciones léxicas WordNet. Aunque no es necesario para realizar esta actividad de laboratorio, WordNet se puede importar utilizando el siguiente comando:

> *from nltk.corpus import wordnet*

In [210]:
# Cargar las instancias de "serve"
instances = senseval.instances('serve.pos')

# Imprimir las primeras 5 instancias con formato amigable
for i, instance in enumerate(instances[:5]):
    print(f"\nInstancia {i + 1}")

    # El contexto es una lista de tuplas (palabra, etiqueta gramatical)
    contexto = ' '.join([w for (w, _) in instance.context])
    print(f"Contexto: {contexto}")

    # Palabra objetivo
    print(f"Palabra objetivo: {instance.word}")

    # Sentidos
    print(f"Sentidos: {', '.join(instance.senses)}")

    print("-" * 50)



Instancia 1
Contexto: some tart fruits mixed with greens make a nice contrast with rich meat dishes ( see orange and onion salad , page 111 ) , but if you like to follow the meat course with sweet fruit , it seems wiser to serve it plain with a good sharp cheese and let it take the place of a sweet or dessert course . if you insist on serving fruit as a salad , don 't cut it into cubes and mix it up .
Palabra objetivo: serve-v
Sentidos: SERVE10
--------------------------------------------------

Instancia 2
Contexto: to increase amount of stuffing , preserve ratio of half as much mushrooms as bread . capon stuffed in this manner would most likely be served with pan-fried potatoes , broccoli , or cauliflower browned in oil .
Palabra objetivo: serve-v
Sentidos: SERVE10
--------------------------------------------------

Instancia 3
Contexto: add a pinch of mace if desired . serve hot or chilled . serves 6 to 8 .
Palabra objetivo: serve-v
Sentidos: SERVE10
-------------------------------

En este laboratorio vas a trabajar con algoritmos basados en aprendizaje automático supervisado, por lo tanto, vas a tener que entrenar diferentes clasificadores que permitan desambiguar las palabras ambiguas en inglés «hard» y «serve», y vas a tener que evaluar el desempeño de los clasificadores creados.

Las diferentes partes que forman este laboratorio se indican a continuación.

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Parte 1: Análisis del corpus</span>

Analiza el corpus Senseval 2 que vas a utilizar para entrenar los clasificadores. Para realizar el análisis utiliza las funcionalidades que aporta NLTK. Desarrolla el código necesario y responde a las siguientes preguntas.

* ¿Cuántos posibles sentidos tienen las palabras ambiguas «hard» y «serve»? ¿Cuáles son esos sentidos? Para cada sentido indica la etiqueta que aparece en el corpus.

Método que a partir de la palabra ambigua, o mejor dicho del nombre del archivo del corpus `word` para la palabra ambigua, obtiene cuáles son los posibles **sentidos de la palabra ambigua**:

In [211]:
def senses(word):
    """
    This takes a target word from senseval-2 and it returns the list of possible senses for the word
    """
    return list(set(
        s
        for instance in senseval.instances(word)
        for s in instance.senses
    ))


Mostrar los sentidos de las palabras ambiguas:

In [212]:
senses('hard.pos')

['HARD1', 'HARD3', 'HARD2']

In [213]:
senses('serve.pos')

['SERVE6', 'SERVE12', 'SERVE10', 'SERVE2']

In [214]:
from nltk.corpus import senseval
from collections import defaultdict

def analizar_sentidos(palabra):
    instancias = senseval.instances(palabra)
    conteo = defaultdict(int)
    ejemplos = {}

    for instancia in instancias:
        for sentido in instancia.senses:
            conteo[sentido] += 1
            if sentido not in ejemplos:
                contexto = ' '.join([w for (w, _) in instancia.context])
                ejemplos[sentido] = contexto

    return conteo, ejemplos

def imprimir_tabla_analisis(palabra):
    conteo, ejemplos = analizar_sentidos(palabra)
    print(f"\n{'='*30} {palabra} {'='*30}")
    print(f"{'Sentido':<10} | {'# de instancias':<15} | {'Ejemplo de contexto':<50}")
    print("-"*100)
    for sentido in sorted(conteo):
        ejemplo = ejemplos[sentido][:50]
        print(f"{sentido:<10} | {conteo[sentido]:<15} | {ejemplo}")
    print("-"*100)

# Ejecutar para hard y serve
imprimir_tabla_analisis('hard.pos')
imprimir_tabla_analisis('serve.pos')



============================== hard.pos ==============================
Sentido    | # de instancias | Ejemplo de contexto                               
----------------------------------------------------------------------------------------------------
HARD1      | 3455            | `` he may lose all popular support , but someone h
HARD2      | 502             | keep this one in your drawer for the next time the
HARD3      | 376             | my companion enjoyed a healthy slice of the chocol
----------------------------------------------------------------------------------------------------

============================== serve.pos ==============================
Sentido    | # de instancias | Ejemplo de contexto                               
----------------------------------------------------------------------------------------------------
SERVE10    | 1814            | some tart fruits mixed with greens make a nice con
SERVE12    | 1272            | mfs multimarket income trust 

* ¿Cuántas instancias hay en el corpus para cada uno de los sentidos de las palabras ambiguas «hard» y «serve»? Es decir, cuantas oraciones hay en el corpus etiquetadas con cada uno de los sentidos.

Método que obtiene todas las **instancias de un posible sentido de la palabra ambigua** a partir de la lista de instancias de una palabra `instances` y el nombre del sentido palabra ambigua `sense`:

In [215]:
def sense_instances(instances, sense):
    """
    This returns the list of instances in instances that have the sense `sense`
    """
    return [
        instancia
        for instancia in instances
        if sense in instancia.senses
    ]

In [216]:
# Obtener las instancias del corpus
hard_instances = senseval.instances('hard.pos')
serve_instances = senseval.instances('serve.pos')

# Ejemplo: obtener todas las instancias con sentido HARD1
hard1 = sense_instances(hard_instances, 'HARD1')
serve6 = sense_instances(serve_instances, 'SERVE6')

# Mostrar cuántas instancias hay
print(f"Instancias con sentido HARD1: {len(hard1)}")
print(f"Instancias con sentido SERVE6: {len(serve6)}")


Instancias con sentido HARD1: 3455
Instancias con sentido SERVE6: 439


* En el contexto, las palabras ambiguas pueden aparecer en diferentes formas gramaticales. Por ejemplo, en el caso de la palabra ambigua *«hard»*, esta aparece tanto la forma base, el adjetivo *«hard»* como en comparativo *«harder»* y como en superlativo *«hardest»*. ¿Qué formas gramaticales aparecen en el contexto para cada una de las palabas ambiguas *«hard»* y *«serve»*?

In [309]:
def formas_gramaticales(instances, base_word):
    formas = set()
    for instance in instances:
        for elemento in instance.context:
            word = elemento[0] if isinstance(elemento, tuple) else elemento
            if word.lower().startswith(base_word.lower()):
                formas.add(word.lower())
    return sorted(formas)



In [310]:
formas_hard = formas_gramaticales(hard_instances, "hard")
formas_serve = formas_gramaticales(serve_instances, "serve")

In [311]:
print("Formas gramaticales reales de 'hard':", formas_hard)
print("Formas gramaticales reales de 'serve':", formas_serve)

Formas gramaticales reales de 'hard': ['hard', 'hardaway', 'harder', 'hardest']
Formas gramaticales reales de 'serve': ['serve', 'served', 'server', 'serves']


Método que muestra las diferentes **versiones de la palabra ambigua** a partir del nombre del archivo del corpus `word` para la palabra ambiguala palabra ambigua:

In [220]:
def tokens(word):
    """
    This takes a target word from senseval-2 and it returns the list of possible tokens for the word
    """
    token = set()
    for instance in senseval.instances(word):
        for elemento in instance.context:
            palabra = elemento[0] if isinstance(elemento, tuple) else elemento
            if palabra.lower().startswith(word.split('.')[0]):
                token.add(palabra.lower())
    return list(token)

Mostrar las diferentes versiones de la palabra ambigua

In [314]:
tokens('hard.pos')

['hardaway', 'hardest', 'hard', 'harder']

In [313]:
tokens('serve.pos')

['served', 'server', 'serve', 'serves']

* ¿Tienen todas las instancias que forman el corpus el formato que se ha descrito anteriormente? Si hay alguna instancia que no cumpla con ese formato, indica cuales serían las incongruencias que presenta y muestra algunos ejemplos.

In [223]:
def verificar_formato_contexto(instances):
    errores = []
    for i, instancia in enumerate(instances):
        for elem in instancia.context:
            if not isinstance(elem, tuple) or len(elem) != 2:
                errores.append((i, elem))
                break  # Ya con una mal en la instancia basta
    return errores

# Revisamos ambas palabras
errores_hard = verificar_formato_contexto(senseval.instances('hard.pos'))
errores_serve = verificar_formato_contexto(senseval.instances('serve.pos'))

# Mostrar errores encontrados
print(f"Errores en 'hard.pos': {len(errores_hard)} instancias mal formateadas")
for i, ejemplo in errores_hard[:3]:  # solo muestra 3 ejemplos
    print(f"- Instancia {i}: {ejemplo}")

print(f"\nErrores en 'serve.pos': {len(errores_serve)} instancias mal formateadas")
for i, ejemplo in errores_serve[:3]:
    print(f"- Instancia {i}: {ejemplo}")

Errores en 'hard.pos': 19 instancias mal formateadas
- Instancia 21: FRASL
- Instancia 110: FRASL
- Instancia 196: FRASL

Errores en 'serve.pos': 76 instancias mal formateadas
- Instancia 18: FRASL
- Instancia 24: FRASL
- Instancia 133: FRASL


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Parte 2: Extracción de características</span>

Para poder entrenar un clasificador es necesario extraer un conjunto de características lingüísticas a partir del corpus etiquetado. Por lo tanto, debes crear el código en Python que te permita extraer diferentes conjuntos de características a partir de Senseval 2.

### Extracción de características basada en las palabras vecinas

Debes extraer un **conjunto de características basado en las palabras vecinas**. Para una instancia del corpus, debes desarrollar el código que sea capaz de extraer el vector de características que indican si las palabras de un vocabulario, que se debe construir previamente, aparecen o no en el contexto (la oración completa en que aparece la palabra ambigua).


Para una instancia de la palaba ambigua «hard» su contexto se muestra a continuación:

In [224]:
instances_hard = senseval.instances('hard.pos')

In [225]:
inst_hard_1 = instances_hard[1]

In [226]:
from tabulate import tabulate

# Obtener la instancia
inst_hard_1 = senseval.instances('hard.pos')[1]

# Separar palabra y etiqueta si es tupla
contexto = [
    (w[0], w[1]) if isinstance(w, tuple) else (w, '')
    for w in inst_hard_1.context
]

# Crear tabla con encabezados
tabla = tabulate(contexto, headers=["Palabra", "Etiqueta"], tablefmt="grid")

# Mostrar
print(f"Sentido de la instancia: {inst_hard_1.senses[0]}")
print(tabla)

Sentido de la instancia: HARD1
+---------------+------------+
| Palabra       | Etiqueta   |
+===============+============+
| clever        | NNP        |
+---------------+------------+
| white         | NNP        |
+---------------+------------+
| house         | NNP        |
+---------------+------------+
| ``            | ``         |
+---------------+------------+
| spin          | VB         |
+---------------+------------+
| doctors       | NNS        |
+---------------+------------+
| ''            | ''         |
+---------------+------------+
| are           | VBP        |
+---------------+------------+
| having        | VBG        |
+---------------+------------+
| a             | DT         |
+---------------+------------+
| hard          | JJ         |
+---------------+------------+
| time          | NN         |
+---------------+------------+
| helping       | VBG        |
+---------------+------------+
| president     | NNP        |
+---------------+------------+
| bush  

Suponiendo que el vocabulario usado para crear extraer las características es el siguiente:

`['time', 'would', 'get', 'work', 'find', 'make']`

Entonces el vector de características extraídas para esa instancia sería:

`{'contains(time)': True,
'contains(would)': False,
'contains(get)': False,
'contains(work)': False,
'contains(find)': False,
'contains(make)': False}`

Este vector de características indica que en el contexto de la palabra ambigua aparece la palabra *«time»* y no aparecen las palabras *«would»*, *«get»*, *«work»*, *«find»* y *«make»*.

Para extraer el vector de características basado en las palabras vecinas debes seguir los siguientes pasos:

#### 1. Construcción del vocabulario o bags of words.

Como ya se ha indicado, para poder obtener el vector de características, se debe construir previamente un vocabulario. Para cada una de las palabras del vocabulario, se debe consultar si la palabra aparece en el contexto de la palabra ambigua. Si la palabra del vocabulario aparece en el contexto entonces en el vector de características aparecerá True para esa palabra y si no, False.

Tomemos como ejemplo un vocabulario creado sobre el que se ha construido el vector de características. Este vocabulario es `['time', 'would', 'get', 'work', 'find', 'make']`. Este vocabulario se usará posteriormente para construir el vector de características del ejemplo que usamos para darte una orientación.

- **¿Cómo construyo mi vocabulario o bags of words?** Lo que debes hacer para tu entrega de este laboratorio es utilizar como vocabulario las m palabras más frecuentes que aparecen en las instancias que conforman el conjunto de datos, es decir en las oraciones que contienen las palabras ambiguas y que forman parte del corpus. Entonces, para crear la **bag of words** (bolsa de palabras) debes extraer el conjunto de las n palabras más frecuentes. Para ello te puedes ayudar de la función `nltk.FreqDist()` que proporciona información sobre la distribución de frecuencias de las palabras que aparecen en un texto.

- Cuando obtengas las palabras más frecuentes, debes eliminar los signos que puntuación y las palabras vacías (aquellas sin significado como artículos, pronombres o preposiciones, las llamadas stop words en inglés). También debes eliminar las diferentes formas gramaticales de la palabra ambigua, por ejemplo, para desambiguar la palabra *«hard»* no tendría sentido utilizar la palabra *«harder»* ni la palabra *«hardest»*.

- Para la eliminación del conjunto de palabras no útiles del vocabulario se puede usar un código parecido al que se indica a continuación. Debes tener en cuenta que **en el código faltaría añadir las palabras que has identificado en la Parte 1** de este laboratorio como las diferentes formas gramaticales de las palabras ambiguas.

In [227]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [228]:
import string

OTHER_WORDS = ["''", '"', "'d", "'ll", "'m", "'re", "'s", "'t", "'ve", '--', '000',
               '1', '2', '3', '4', '5', '6', '8', '10', '15']

STOPWORDS_SET = set(stopwords.words('english')).union(set(string.punctuation), set(OTHER_WORDS))


- Ejemplo de vocabulario para un tamaño de 6. Por ejemplo, si se quiere entrenar un clasificador que permita identificar los diferentes sentidos de la palabra «hard» y se utilizan para entrenar y validar el modelo las instancias etiquetadas para esta palabra, la bolsa de palabras en el caso de considerar las seis palabras más frecuentes (m=6) sería la presentada anteriormente `['time', 'would', 'get', 'work', 'find', 'make']`.

Método que devuelve el **vocabulario** formado por las `m` palabras más frecuentes en el contexto de una palabra ambigua a partir de un conjunto de instancias `instances` y el conjunto de palabras no útiles `stopwords` (Nota: Recuerda que se deben eliminar las diferentes formas gramaticales de la palabra ambigua):

In [229]:
def extract_vocab(instances, stopwords=STOPWORDS_SET, m=250):
    """
    Given a list of senseval instances, return a list of the m most frequent words that
    appears in its context (i.e., the sentence with the target word in), output is in order
    of frequency.
    """
    return [w for w,f in extract_vocab_frequency(instances,stopwords,m)]

In [230]:
def extract_vocab_frequency(instances, stopwords=STOPWORDS_SET, m=250):
    """
    Dada una lista de instancias de senseval, devuelve una lista de las m palabras más frecuentes
    (excluyendo stopwords y formas gramaticales de la palabra ambigua).
    El resultado es una lista de tuplas (palabra, frecuencia), ordenada de mayor a menor frecuencia.
    """
    frecuencia = {}

    for instance in instances:
        palabras_en_contexto = set()  # usamos set para contar solo una vez por instancia
        for elemento in instance.context:
            palabra = elemento[0] if isinstance(elemento, tuple) else elemento
            palabra = palabra.lower()
            if palabra.isalpha() and palabra not in stopwords:
                palabras_en_contexto.add(palabra)

        for palabra in palabras_en_contexto:
            frecuencia[palabra] = frecuencia.get(palabra, 0) + 1

    # Ordenamos por frecuencia descendente
    vocab_ordenado = sorted(frecuencia.items(), key=lambda x: x[1], reverse=True)

    # Retornamos solo los m más frecuentes
    return vocab_ordenado[:m]

Mostrar el vocabulario de tamaño de 6 para la palabra «hard», es decir el vocabulario creado usando las instancias de la a

In [231]:
vocab_6 = extract_vocab(instances_hard, STOPWORDS_SET, m=6)

In [232]:
vocab_6

['hard', 'said', 'harder', 'time', 'would', 'get']

#### 2. Construcción del conjunto de características basado en palabras vecinas.

Utiliza un diccionario en Python para guardar el conjunto de características. La clave del diccionario serán las palabras del vocabulario y el valor debe ser un booleano para indicar la aparición o no de las palabras en el contexto. Por ejemplo, en el vector de características `{'contains(time)': True, 'contains(would)': False, 'contains(get)': False, 'contains(work)': False, 'contains(find)': False, 'contains(make)': False}` una de las claves del diccionario es `'contains(time)'` y su valor es `True`  lo que indica que en el contexto de la palabra ambigua aparece la palabra *«time»*.

- Debes mostrar el vector de características resultante para una de las instancias del corpus.

- **Importante:** En el cómputo del vector de características basado en las palabras vecinas debes utilizar como contexto la oración completa donde aparece la palabra ambigua. Es decir, todas las palabras que forman la oración guardada en el campo  `context` de la instancia.

Método que devuelve el conjunto de características basado en palabras vecinas para una instancia `instance` a partir de un vocabulario `vocab` (Nota: el parámetro `dist` no debes usarlo):

In [233]:
def wsd_caracteristicas_palabras_vecinas(instance, stopwords_set=STOPWORDS_SET, number=250, distance=None):
    features = {}

    context = [w[0].lower() if isinstance(w, tuple) else w.lower() for w in instance.context]

    # Crear lista de palabras contextuales sin stopwords
    content_words = [w for w in context if w not in stopwords_set]

    # Seleccionar las n palabras más frecuentes
    for w in content_words[:number]:
        features[f'contains({w})'] = True

    return features


Mostrar el vector de características basado en palabras vecinas para una de las instancias del corpus usando el vocabulario de seis palabras calculado previamente:

In [234]:
wsd_caracteristicas_palabras_vecinas(inst_hard_1, vocab_6, 0)

{}

### Extracción de características basada en características de colocación

Debes extraer también un **conjunto de características de colocación**. Para una instancia del corpus, debes desarrollar el código que sea capaz de extraer el vector de características formado por la secuencia de n palabras que ocurren antes de la palabra ambigua y la secuencia de n palabras que ocurren después de la palabra ambigua, los llamados n-gramas.

Para una instancia de la palaba ambigua «hard» su contexto se muestra a continuación:

In [235]:
inst_hard_2737 = instances_hard[2737]

In [236]:
from tabulate import tabulate

# Obtener la instancia
inst_hard_2737 = senseval.instances('hard.pos')[2737]

# Separar palabra y etiqueta (si es tupla) o dejarla como string
contexto = [
    (w[0], w[1]) if isinstance(w, tuple) else (w, '')
    for w in inst_hard_2737.context
]

# Crear tabla con encabezado
tabla = tabulate(contexto, headers=["Palabra", "Etiqueta"], tablefmt="grid")

# Mostrar
# Accessing the senses attribute, which is a tuple
print(f"Sentido de la instancia: {inst_hard_2737.senses[0]}")
print(tabla)

Sentido de la instancia: HARD1
+-------------+------------+
| Palabra     | Etiqueta   |
+=============+============+
| ``          | ``         |
+-------------+------------+
| it          | PRP        |
+-------------+------------+
| 's          | VBZ        |
+-------------+------------+
| a           | DT         |
+-------------+------------+
| very        | RB         |
+-------------+------------+
| interesting | JJ         |
+-------------+------------+
| place       | NN         |
+-------------+------------+
| to          | TO         |
+-------------+------------+
| work        | VB         |
+-------------+------------+
| ,           | ,          |
+-------------+------------+
| but         | CC         |
+-------------+------------+
| i           | PRP        |
+-------------+------------+
| can         | MD         |
+-------------+------------+
| see         | VB         |
+-------------+------------+
| why         | WRB        |
+-------------+------------+
| some      

Entonces el vector de características de colocación para el bigrama anterior y posterior sería:

`{'previous(have a)': True, 'next(time imagining)': True}`

Este vector de características indica que antes de la palabra ambigua se encuentran las palabras *«have a»* y después de la palabra ambigua las palabras *«time imagining»*.

- **¿Cómo construyo mi conjunto y que n utilizo?** Utiliza un diccionario en Python para guardar el conjunto de características, la clave del diccionario debe indicar la secuencia de palabras de contexto y si aparecen antes o después de la palabra ambigua y el valor asociado a la clave debe ser un booleano verdadero. Usaremos **n=2**.

- Por ejemplo, en el vector de características `{'previous(have a)': True, 'next(time imagining)': True}` una de las claves del diccionario es `'previous(have a)'` y su valor es `True` lo que indica que antes de la palabra ambigua se encuentran las palabras *«have a»*. En este caso, al tener secuencias de dos palabras (*n=2*), se están considerando bigramas y la ventana tendría tamaño cinco (*2n+1*). Por lo tanto, si la palabra ambigua es *«hard» en el contexto guardado en el campo context de la instancia, aparece la siguiente parte de la frase «have a hard time imagining».

- Debes mostrar el vector de características resultante para una de las instancias del corpus.

- **Importante:** Debes tener en cuenta los posibles casos en los que la palabra ambigua aparezca al principio o final de la frase, ya que en esas instancias no vas a poder obtener una secuencia de palabras de longitud n. Por ejemplo, para la instancia cuyo contexto es: `[('some', 'DT'), ('hard', 'JJ'), ('choices', 'NNS'), ('had', 'VBD'), ('to', 'TO'), ('be', 'VB'), ('made', 'VBN'), …]` si n=2 deberías obtener el siguiente vector de características: `{'previous(some)': True, 'next(choices had)': True}`.

- **Nota:** aunque no es imprescindible para realizar esta actividad de laboratorio, puedes utilizar las funcionalidades para trabajar con n-gramas que ofrece NLTK. Estas se pueden importar utilizando el siguiente comando:

> *from nltk import ngrams*

Método que devuelve el conjunto de características de colocación para una instancia `instance` usando los n-gramas anterior y posterior donde la longitud de secuencia es de `dist` palabras (Nota: el parámetro `vocab` no debes usarlo):

In [264]:
def wsd_caracteristicas_colocacion(instance, vocab=None, dist=2):
    features = {}

    # Extraer contexto en minúsculas
    context = [w.lower() if isinstance(w, str) else w[0].lower() for w in instance.context]

    # Buscar manualmente la palabra ambigua que empiece con "hard"
    target_word = next((w for w in context if w.startswith("hard")), None)

    if target_word is None:
        print(f"No se encontró ninguna palabra que comience con 'hard' en el contexto: {context}")
        return features

    # Buscar índice
    target_index = None
    for i, word in enumerate(context):
        if word == target_word:
            target_index = i
            break

    if target_index is None:
        print(f"No se encontró la palabra ambigua '{target_word}' en el contexto: {context}")
        return features

    # Palabras anteriores y posteriores
    prev_words = context[max(0, target_index - dist):target_index]
    next_words = context[target_index + 1:target_index + 1 + dist]

    # Crear bigramas anteriores
    for i in range(len(prev_words)):
        bigram = '_'.join(prev_words[i:])
        features[f'previous({bigram})'] = True

    # Crear bigramas posteriores
    for i in range(len(next_words)):
        bigram = '_'.join(next_words[:i+1])
        features[f'next({bigram})'] = True

    return features


Mostrar el vector de características de colocación para una de las instancias del corpus usando n=2:

In [265]:
wsd_caracteristicas_colocacion(inst_hard_2737, dist=2)

{'previous(have_a)': True,
 'previous(a)': True,
 'next(time)': True,
 'next(time_imagining)': True}

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Parte 3: Entrenamiento de clasificadores</span>

Debes entrenar diferentes clasificadores que permitan desambiguar las palabras ambiguas en inglés «hard» y «serve». Además, vas a tener que evaluar el desempeño de los clasificadores creados. Por lo tanto, debes crear el código en Python que te permita entrenar estos clasificadores y evaluarlos.

El tipo de clasificador que vas a utilizar en este laboratorio es el Naive Bayes. Para importar el clasificador y el paquete que te permita evaluar su rendimiento debes utilizar el siguiente comando:

In [266]:
from nltk.classify import accuracy, NaiveBayesClassifier
from nltk import ConfusionMatrix

Una vez hayas importado los paquetes anteriores, para entrenar un clasificador Naïve Bayes puedes usar el comando `NaiveBayesClassifier.train()` y para evaluarlo `accuracy()`. Además, puedes utilizar el clasificador entrenado para clasificar una instancia utilizando su método `classify()`. Por último, puedes obtener la matriz de confusión utilizando el comando `nltk.ConfusionMatrix()`.

Para realizar esta parte del laboratorio debes seguir los siguientes pasos:

### 1.	Entrenamiento de dos clasificadores para la palabra «hard».

Debes entrenar dos clasificadores que permitan desambiguar la palabra «hard», es decir, que debes entrenar los clasificadores utilizando las instancias disponibles en el corpus Senseval 2 para esta palabra ambigua.

- **Conjuntos de entrenamiento y test.** Para entrenar y validar divide las instancias disponibles en una proporción del 80-20 % y recuerda que en el conjunto de datos de entrenamiento deben aparecer instancias de todas las clases.

- **Clasificador basado en las palabras vecinas.** Con el conjunto de datos de entrenamiento, entrena un clasificador para «hard» que use como características el conjunto basado en las **palabras vecinas** cuyo código has implementado en la parte 2 de este laboratorio. Para definir el vocabulario utiliza las **250** palabras más frecuentes (**m=250**).

- **Clasificador basado en características de colocación.** Con el conjunto de datos de entrenamiento, entrena un clasificador para «hard» que use como conjunto de características las de colocación cuyo código has implementado en la parte 2 de este laboratorio. Para definir la **ventana de contexto** utiliza la secuencia de dos palabras que ocurren antes de la palabra ambigua y la secuencia de dos palabras que ocurren después de esta **(n=2)**.

In [285]:
from nltk.classify import accuracy, NaiveBayesClassifier
from nltk import ConfusionMatrix
import random

def wsd_clasificador(word, features_func, stopwords_list=STOPWORDS_SET, number=250, vocab=None, distance=2, errors=False, confusion_matrix=False):

    # Filtrar instancias para la palabra ambigua específica
    instances = [inst for inst in senseval.instances(word)]

    # Mezclar para garantizar aleatoriedad
    random.seed(42)
    random.shuffle(instances)

    # Dividir en 80% train y 20% test
    split_point = int(0.8 * len(instances))
    train_data = instances[:split_point]
    test_data = instances[split_point:]

    # Construir conjuntos de entrenamiento y prueba con features
    if features_func == wsd_caracteristicas_palabras_vecinas:
        train_set = [(features_func(inst, stopwords_list, number), inst.senses[0]) for inst in train_data]
        test_set  = [(features_func(inst, stopwords_list, number), inst.senses[0]) for inst in test_data]
    elif features_func == wsd_caracteristicas_colocacion:
        train_set = [(features_func(inst, vocab, distance), inst.senses[0]) for inst in train_data]
        test_set  = [(features_func(inst, vocab, distance), inst.senses[0]) for inst in test_data]
    else:
        train_set = [(features_func(inst), inst.senses[0]) for inst in train_data]
        test_set  = [(features_func(inst), inst.senses[0]) for inst in test_data]

    # Entrenar el clasificador
    classifier = NaiveBayesClassifier.train(train_set)

    # Evaluar accuracy
    acc = accuracy(classifier, test_set)
    print('Accuracy: %6.4f' % acc)

    # Mostrar errores si se solicita
    if errors:
        print('\nErrores:')
        errores = []
        for i, (featureset, gold) in enumerate(test_set):
            guess = classifier.classify(featureset)
            if guess != gold:
                instance = test_data[i]
                errores.append((instance, guess))

                context_words = [w[0] if isinstance(w, tuple) else w for w in instance.context]
                target_word = instance.word.lower()
                target_index = None
                for j, w in enumerate(context_words):
                    if isinstance(w, tuple):
                        w_clean = w[0].lower()
                    else:
                        w_clean = w.lower()
                    if w_clean == target_word:
                        target_index = j
                        break

                if target_index is not None:
                    start_index = max(0, target_index - 2)
                    end_index = min(len(context_words), target_index + 3)
                    context_snippet = ' '.join(context_words[start_index:end_index])
                    print(f"{i:>2}) {context_snippet}")
                    print(f"   Predicted: {guess}, Actual: {gold}")
                else:
                    print(f"{i:>2}) [Context not found] Predicted: {guess}, Actual: {gold}")

        return errores  # Retornar lista de errores

    # Mostrar matriz de confusión si se solicita
    if confusion_matrix:
        gold_labels = [label for (_, label) in test_set]
        predicted_labels = [classifier.classify(fs) for (fs, _) in test_set]
        cm = ConfusionMatrix(gold_labels, predicted_labels)
        print('\nMatriz de confusión:')
        print(cm)

    return classifier  # Retornar el clasificador si no se solicitaron errores


### 2.	Validación de los clasificadores para la palabra «hard».  

Utilizando el conjunto de datos de test que has generado previamente, obtén la exactitud (accuracy) y la matriz de confusión para cada uno de los dos clasificadores que permiten desambiguar el sentido de la palabra «hard».

- Debes mostrar la exactitud (accuracy) y la matriz de confusión resultantes de la validación de cada uno de los dos clasificadores.

In [286]:
clasificador_hard_vecinas = wsd_clasificador('hard.pos', wsd_caracteristicas_palabras_vecinas, number=250, errors=False, confusion_matrix=True)

Accuracy: 0.5652

Matriz de confusión:
      |   H   H   H |
      |   A   A   A |
      |   R   R   R |
      |   D   D   D |
      |   1   2   3 |
------+-------------+
HARD1 |<355>208 130 |
HARD2 |  14 <82> 16 |
HARD3 |   2   7 <53>|
------+-------------+
(row = reference; col = test)



In [287]:
clasificador_hard_colocacion = wsd_clasificador(
    'hard.pos',
    wsd_caracteristicas_colocacion,
    vocab={},                # Ya no causará error
    distance=2,
    errors=False,
    confusion_matrix=True
)


Accuracy: 0.8881

Matriz de confusión:
      |   H   H   H |
      |   A   A   A |
      |   R   R   R |
      |   D   D   D |
      |   1   2   3 |
------+-------------+
HARD1 |<647> 31  15 |
HARD2 |  18 <84> 10 |
HARD3 |  13  10 <39>|
------+-------------+
(row = reference; col = test)



Debes comparar y analizar los resultados de rendimiento de los clasificadores. Para ello responde a las siguientes preguntas:

- ¿Cuál es el conjunto de características que aporta mejores resultados? ¿Por qué?

El conjunto de características basado en colocación `(wsd_caracteristicas_colocacion)` aporta mejores resultados, alcanzando una exactitud del 88.81%, frente al 56.52% del clasificador basado en palabras vecinas.
Esto se debe a que los bigramas capturan mejor las dependencias locales y patrones lingüísticos típicos del contexto inmediato de la palabra ambigua, lo cual resulta más informativo para el desambiguador que simplemente tomar un número fijo de palabras cercanas sin tener en cuenta su relación estructural.

- ¿Cuál es el sentido más difícil de identificar? ¿Por qué?

El sentido HARD3 es el más difícil de identificar.
Esto se observa en ambas matrices de confusión, donde HARD3 muestra el menor número de aciertos (solo 39 en colocación y 53 en vecinas) y una mayor proporción de errores respecto a su total. Esto probablemente se debe a que este sentido es menos frecuente en el corpus y/o su contexto no tiene patrones tan distintivos como los otros sentidos, lo cual genera ambigüedad para el clasificador.

- ¿Qué posibles mejoras se podrían aplicar para mejorar el rendimiento de los clasificadores creados? No es necesario que las implementes, solo que las comentes.

Algunas posibles mejoras incluyen:

1.   Aumentar el tamaño del corpus de entrenamiento para mejorar la cobertura de contextos y balancear los sentidos menos frecuentes.
2.   Utilizar modelos más complejos como `TF-IDF`, word embeddings `(por ejemplo, Word2Vec, GloVe)` o incluso modelos contextuales como `BERT`.
3.   Incluir información gramatical o sintáctica adicional, como etiquetas POS más amplias o árboles de dependencia.
4.   Aplicar técnicas de data augmentation para balancear clases.
5.  Ajustar los hiperparámetros (como distance o number) mediante validación cruzada.

### 3.	Instancias clasificadas incorrectamente para «hard».   

Para el clasificador que permite desambiguar la palabra «hard» y que utiliza las características de colocación, obtén las instancias que pertenecen al sentido ‘HARD1’ y que se han clasificado incorrectamente. Presenta en el informe la oración en la que aparece la palabra ambigua (el contexto) para cada una de esas instancias y la etiqueta en la que han sido erróneamente clasificadas.

Para el clasificador que permite desambiguar la palabra «hard» y que utiliza las características de colocación, obtén las instancias que pertenecen al sentido ‘HARD1’ y que se han clasificado incorrectamente.

- Presenta la oración en la que aparece la palabra ambigua (el contexto) para cada una de esas instancias y la etiqueta en la que han sido erróneamente clasificadas.

In [295]:
# Ejecutar el clasificador con errors=True y confusion_matrix=False
errores = wsd_clasificador(
    'hard.pos',
    wsd_caracteristicas_colocacion,
    distance=2,
    errors=True,
    confusion_matrix=False
)

Accuracy: 0.8881

Errores:
13) [Context not found] Predicted: HARD2, Actual: HARD1
14) [Context not found] Predicted: HARD1, Actual: HARD3
35) [Context not found] Predicted: HARD1, Actual: HARD3
52) [Context not found] Predicted: HARD2, Actual: HARD1
69) [Context not found] Predicted: HARD3, Actual: HARD1
71) [Context not found] Predicted: HARD1, Actual: HARD3
92) [Context not found] Predicted: HARD2, Actual: HARD3
99) [Context not found] Predicted: HARD1, Actual: HARD2
100) [Context not found] Predicted: HARD2, Actual: HARD1
109) [Context not found] Predicted: HARD2, Actual: HARD3
111) [Context not found] Predicted: HARD3, Actual: HARD2
124) [Context not found] Predicted: HARD3, Actual: HARD2
143) [Context not found] Predicted: HARD1, Actual: HARD2
148) [Context not found] Predicted: HARD1, Actual: HARD3
149) [Context not found] Predicted: HARD3, Actual: HARD1
154) [Context not found] Predicted: HARD2, Actual: HARD1
157) [Context not found] Predicted: HARD1, Actual: HARD3
164) [Contex

In [294]:
print("Errores para instancias HARD1 mal clasificadas por colocación:\n")

for instancia, pred in errores:
    if instancia.senses[0] == 'HARD1' and pred != 'HARD1':
        contexto = " ".join(w[0] if isinstance(w, tuple) else w for w in instancia.context) if instancia.context else "[Contexto no disponible]"
        print("Oración (contexto):")
        print(contexto)
        print(f"Palabra ambigua: {instancia.word}")
        print(f"Etiqueta real: {instancia.senses[0]}, Predicción: {pred}")
        print("-" * 60)


Errores para instancias HARD1 mal clasificadas por colocación:

Oración (contexto):
-- j . d . vanderlaan ; los gatos ; don 't let your pets get spooked oct . 31 ; halloween is around the corner , a hard time for our four-legged friends .
Palabra ambigua: hard-a
Etiqueta real: HARD1, Predicción: HARD2
------------------------------------------------------------
Oración (contexto):
`` it 's been a long , hard fight .
Palabra ambigua: hard-a
Etiqueta real: HARD1, Predicción: HARD2
------------------------------------------------------------
Oración (contexto):
( they ) are living up to their standards , as hard as it is for you to imagine .
Palabra ambigua: hard-a
Etiqueta real: HARD1, Predicción: HARD3
------------------------------------------------------------
Oración (contexto):
but she also said this is abc 's first season of scrambling and that pilfering signals is likely to get harder next year .
Palabra ambigua: hard-a
Etiqueta real: HARD1, Predicción: HARD2
---------------------

### 4.	Entrenamiento y validación de dos clasificadores para la palabra «serve».   

Repite el proceso anterior para entrenar y validar dos clasificadores que permitan desambiguar la palabra «serve». Puedes aprovechar el código que has generado anteriormente.

-	Crea los conjuntos de entrenamiento y test para las instancias donde la palabra ambigua es «serve». Mantén la proporción del 80-20 % para la creación de los conjuntos de entrenamiento y de test.

-	Entrena un clasificador para «serve» que use como características el conjunto basado en las **palabras vecinas**. Para definir el vocabulario utiliza las **250** palabras más frecuentes (**m=250**).

-	Entrena un clasificador para «serve» que use como conjunto **de características las de colocación**. Para definir la **ventana de contexto** utiliza la secuencia de dos palabras que ocurren antes de la palabra ambigua y la secuencia de dos palabras que ocurren después de esta **(n=2)**.

-	Obtén el valor la exactitud (accuracy) para cada uno de los dos clasificadores que permiten desambiguar el sentido de la palabra «sense».

In [299]:
clasificador_serve_vecinas = wsd_clasificador('serve.pos', wsd_caracteristicas_palabras_vecinas, number=250)

Accuracy: 0.7454


In [300]:
clasificador_serve_colocacion = wsd_clasificador('serve.pos', wsd_caracteristicas_colocacion, distance=2, errors=False)

No se encontró ninguna palabra que comience con 'hard' en el contexto: ['transtar', 'airlines', ',', 'a', 'subsidiary', 'of', 'dallas-based', 'southwest', 'airlines', ',', 'said', 'it', 'will', 'form', 'a', 'commuter', 'airline', 'with', 'closely', 'held', 'rio', 'airways', 'to', 'serve', 'six', 'texas', 'cities', '.', 'the', 'service', ',', 'which', 'goes', 'into', 'effect', 'jan', '.', '7', ',', 'will', 'serve', 'brownsville-south', 'padre', 'island', ',', 'bryan-college', 'station', ',', 'victoria', ',', 'laredo', ',', 'san', 'angelo', 'and', 'killeen', '.']
No se encontró ninguna palabra que comience con 'hard' en el contexto: ['some', 'valuable', 'characters', 'are', 'about', 'to', 'be', 'lost', 'to', 'the', 'house', 'of', 'representatives', 'of', 'their', 'own', 'choice', '.', 'i', 'feared', 'once', 'that', 'this', 'would', 'be', 'the', 'case', 'with', 'mr', '.', 'smith', ',', 'of', 'your', 'state', ';', 'but', 'i', 'believe', 'his', 'present', 'intention', 'is', 'rather', 'to', 

### 5.	Análisis de resultados del rendimiento de los clasificadores y conclusiones sobre el uso de aprendizaje automático supervisado para desambiguar el sentido de las palabras

Presenta una tabla resumen con los valores de exactitud para cada uno de los 4 clasificadores (dos para cada palabra ambigua) que has entrenado previamente y responde a las siguientes preguntas:

########## Aquí debes indicar tu respuesta ##########

* ¿Por qué no es justo comparar directamente la exactitud aportada por los clasificadores que han aprendido diferentes palabras ambiguas?

########## Aquí debes indicar tu respuesta ##########

* ¿Cómo podrías hacerlo para que la comparación entre clasificadores que desambiguan palabras diferentes tenga sentido?

########## Aquí debes indicar tu respuesta ##########

* Compara la exactitud de los clasificadores con la que proporcionaría un clasificador que asignara el sentido de forma aleatoria. ¿Cuál sería el mejor clasificador tomando como referencia (baseline), el clasificador aleatorio?

########## Aquí debes indicar tu respuesta ##########

Una vez hayas implementado diferentes clasificadores para desambiguar el sentido de diferentes palabras y analizado su desempeño, reflexiona sobre el uso de algoritmos basados en aprendizaje automático supervisado para resolver la tarea de desambiguación del sentido de las palabras. Para ello responde de forma razonada a las siguientes preguntas:

* ¿Cuáles son las limitaciones de los clasificadores que has creado para la desambiguación del sentido de las palabras?

########## Aquí debes indicar tu respuesta ##########

* ¿Qué alternativas propondrías para superar esas limitaciones y obtener algoritmo que resuelva mejor el problema de la desambiguación del sentido de las palabras?

########## Aquí debes indicar tu respuesta ##########

I've added a cell to download the necessary NLTK data. Now I will modify the code cell to correctly import and use the stopwords.